src/pseudo/PspUpf.jl

```diff
@@ -0,0 +1,220 @@
+    length(pswfcs) == length(pswfc_ang_moms) || error("Length of pseudo wfcs and pseudo" *
+                                                      " wfc angular momenta do not agree.")
+    length(pswfcs) == length(pswfc_occs)     || error("Length of pseudo wfcs and pseudo" *
+                                                      " wfc occupations do not agree.")
+
+    return PspUpf{T,IT}(Zion, lmax, rgrid, drgrid, vloc, vloc_interp, projs, projs_interp, h,
+                     pswfcs, pswfc_ang_moms, pswfc_occs, rhoatom, rvlocpzdr, r2projsdr,
+                     identifier)
+end
+
+function parse_upf_file(path; identifier=path)
+    pseudo = load_upf(path)
+
+    unsupported = []
+    pseudo["header"]["core_correction"]      && push!(unsupported, "non-lin. core correction")
+    pseudo["header"]["has_so"]               && push!(unsupported, "spin-orbit coupling")
+    pseudo["header"]["pseudo_type"] == "SL"  && push!(unsupported, "semilocal potential")
+    pseudo["header"]["pseudo_type"] == "US"  && push!(unsupported, "ultrasoft")
+    pseudo["header"]["pseudo_type"] == "PAW" && push!(unsupported, "plane-augmented wave")
+    pseudo["header"]["has_gipaw"]            && push!(unsupported, "gipaw data")
+    pseudo["header"]["pseudo_type"] == "1/r" && push!(unsupported, "Coulomb")
+    length(unsupported) > 0 && error("Pseudopotential contains the following unsupported" *
+                                     " features/quantities: $unsupported")
+
+    Zion        = Int(pseudo["header"]["z_valence"])    # Pseudo-ion charge
+    rgrid       = pseudo["radial_grid"]                 # Radial grid
+    drgrid      = pseudo["radial_grid_derivative"]      # Integration coeff.s
+    lmax        = pseudo["header"]["l_max"]             # Maximum angular momentum channel
+    nproj       = pseudo["header"]["number_of_proj"]    # Number of BKB projectors
+    nwfc        = pseudo["header"]["number_of_wfc"]     # Number of pseudo-atomic ψ
+    vloc        = pseudo["local_potential"] ./ 2        # Local potential (Ry -> Ha)
+    vloc_interp = linear_interpolation((rgrid,), vloc)  # Interp.or for the local potential
+    IT = typeof(vloc_interp)
+
+    # Kleinman-Bylander β projectors projs[l+1][i][ir]
+    # NB: UPFs store rβ, not just β
+    projs = [Vector[] for _ = 0:lmax]   # Projectors on the grid
+    projs_interp = [IT[] for _ = 0:lmax]  # Interpolators for the non-local projectors
+    for i = 1:nproj
+        proj_data = pseudo["beta_projectors"][i]
+        l = proj_data["angular_momentum"]
+        proj = proj_data["radial_function"] ./ 2  # Ry -> Ha
+        ir_start = iszero(rgrid[1]) ? 2 : 1  # Some grids start at 0., so (rβ)[1]/r[1] is undefined
+        ir_cut = length(proj)  # Projectors are cut off by the UPF parser @ ir_cut given by the UPF
+        proj_interp = linear_interpolation((rgrid[ir_start:ir_cut],), proj[ir_start:ir_cut])
+        push!(projs[l+1], proj)
+        push!(projs_interp[l+1], proj_interp)
+    end
```

Comment:
I would use a `map`-based construct (over `pseudo["beta_projectors"]`) to avoid the need to explicitly deduce the type of `projs_interp` and `projs`. It's not that `nproj` is super large, so running over all of them twice should be ok.